pkg/metrics/retries.go

```diff
@@ -0,0 +1,74 @@
+package metrics
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	// Retries observes finished retried operations.
+	Retries RetriesMetric = &retriesMetric{}
+)
+
+func init() {
+	Retries.(*retriesMetric).init()
+}
+
+// RetriesMetric observes finished retried operations.
+type RetriesMetric interface {
+	// Observe performs a single observation of a finished retry loop.
+	// codeLocation is a string representation of the code location
+	// of the retry loop.
+	// retryCount is the number of retries performed, where the very
+	// first attempt is not counted as retry.
+	// latency is the elapsed time from the start of the retry loop
+	// until is has been finished.
+	Observe(codeLocation string, retryCount uint64, latency time.Duration)
+}
+
+type retriesMetric struct {
+	initOnlyOnce  sync.Once
+	countMetric   *prometheus.HistogramVec
+	latencyMetric *prometheus.HistogramVec
+}
+
+func (m *retriesMetric) init() {
+	m.initOnlyOnce.Do(func() {
+		m.countMetric = prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Subsystem: Subsystem,
+				Name:      "retry_count",
+				Help:      "Generic metric for retry loops collecting the number of retries performed for retried operations.",
+				// TODO review buckets
+				Buckets: prometheus.ExponentialBuckets(1, 2, 32),
```

Comment:
```suggestion
				Buckets: append(prometheus.LinearBuckets (0, 1, 10),prometheus.LinearBuckets (10, 5, 10),
```

Reply:
Given the fact that this is a _generic_ metric, there might measured retry loops that typically go beyond 75 retries. On the other hand we should not have too many buckets to not occupy too much space in the metrics store.

Here's my proposal:

```suggestion
				// TODO review buckets
				Buckets: append(prometheus.LinearBuckets (0, 1, 10), prometheus.ExponentialBuckets(10, 2, 11)),
```

This gives the following bucket upper bounds:

1
2
3
4
5
6
7
8
9
10
20
40
80
160
320
640
1280
2560
5120
10240